### DISCUSSION TITLE
PipeWire support

### DISCUSSION THREAD

**Discussion by columbarius** [obsproject/obs-studio#7998] [General]
Summary of the current situation:

Small improvements for screencapture:

- Improve detection of invalid buffers: #5614
- Handle buffer transformation on import (rotations and flips): #7762

Small bug fixes:
- Not crash on software rendering: #7986 

Big features:

- Uncouple the PipeWire code from screencast so we can reuse it more easily for other features like camera, maybe audio (contains lot of refactoring/cleanups) #6669 (this blocks the following video related MRs)
- Virtual camera #5377 
- Webcam/Device capture #5728
- Audio capture #6207 

Problems of the current situation:

The unified PipeWire wrapper collects a lot of cleanup and small fixes together with design decisions influencing the way forward for the other feature MRs. This slows down the feature MR since it is not certain on what state they should be based on. Additionally this catch all approach makes it annoying to review. 

Ways forward:

- Move a lot of those cleanup commits into small MR's or a series, which can be merged easier and quicker, making the review of the wrapper simpler. Here the quicker is important to get the wrapper changes close to master again without having it depend now on a list of smaller commits.
- Don't base the new features (audio, device, vcam) on the wrapper idea, but duplicate the core PipeWire code and then unify it later. This unblocks those immediately.

Caveats

- All of those big video related MR's are currently in a not well defined/rough state. They should be rebased and I might have different versions of them depending on integration with the wrapper idea. 
- The device capture has some jumps/frame drops. To debug them we probably need another client implementation to compare.

**Comment by kkartaltepe** [comment]
I'm happy to review a large collection of small fixup MRs (preferably as few PRs as makes sense, e.g. a bunch of one line fixup commits can be one PR but larger fixes might need their own), so we can merge those sooner. But I would assume Georges might prefer to review the feature/subsystem merge work.

**Comment by vkedwardli** [comment]
Hi, I'm not familiar with the PipeWire in Linux env, is it similar to Syphon (Mac) or Spout (Windows), which can route the video directly using GPU texture? (DMA-BUF?)

Seems that the Desktop Capture or Window Capture is using PipeWire, but is there a direct way to route the GPU texture into OBS using PipeWire?

**Reply by kkartaltepe** [reply]
Pipewire provides the sinks and sources, and presumably some audio routers could just do the routing for video too. But no one has built any such tools or defined any protocols to coordinate between video tools like that (ala Jack). And im not aware of anyone working on that.

I think there are thoughts that implementing camera capture can functionally serve as the protocol for general program to program sharing of video. Where programs just pretend they are "cameras". And there are people working on making cameras work on pipewire.

**Reply by pallaswept** [reply]
> no one has built any such tools

There's [qpwgraph](https://gitlab.freedesktop.org/rncbc/qpwgraph) and [helvum](https://gitlab.freedesktop.org/pipewire/helvum) based on qjackctl and catia respectively (ala Jack)

> is there a direct way to route the GPU texture into OBS using PipeWire?

Yes. https://docs.pipewire.org/page_dma_buf.html

HTH

**Reply by vkedwardli** [reply]
In Windows, we can use https://github.com/Off-World-Live/obs-spout2-plugin
In Mac, Syphon support is built-in in OBS

@pallaswept it is possible to use DMA-BUF Sharing in pipewire, but how to pipe it into OBS?

**Reply by pallaswept** [reply]
> how to pipe it into OBS?

I'm not quite sure what you're asking here.

The idea of DMA-BUF sharing is that you don't copy the buffer around, you modify it in place on the GPU's VRAM. This way, you avoid using CPU cycles to control VRAM to DRAM and back to VRAM transfers across the PCI buss, which frees those resources and allows you to manipulate the data more efficiently. You might be familiar with this kind of approach as "nvenc new' or 'Jim nvenc' in OBS. The above link explains how that's done. via pipewire. There's no simple explanation beyond that link, it's rocket science stuff.

If you want to copy the content of the buffer to OBS, you write a pipewire client to capture the video stream, which is what we have now already with the pipewire screen and window capture features.

**Reply by vkedwardli** [reply]
May be you can take a look at what we can do in Mac / Windows first: https://github.com/flyinghead/flycast/pull/1126

I know pipewire supports sharing DMA-BUF, I can also modify my program to share the DMA-BUF, but I don't understand how can I actually display the shared DMA-BUF in OBS on Linux in the current Pipewire screen/window capture implementation, since i don't need to capture, but instead I'm sharing directly from my program?

**Reply by pallaswept** [reply]
I just left Windows on my home PC since I work with linux and prefer it and it's almost ready for the desktop. I'm familiar with how Windows works. In linux land the kernel shares the DMA buffers. All the details you need are in the link I shared with you and the link you just shared with me.

**Reply by vkedwardli** [reply]
I'm not trying to implement the relevant part on OBS, I know it is possible to implement sharing DMA-BUF in my program, I'm posting this question to ask if there is already a way to accept this shared buffer directly in the current implementation of OBS.

What confuses me is that, OBS seems to be support pipewire already, but only limited to Desktop/Window capture.
So I'm posting this question and hoping someone would clarify on:
"If it is possible to use the current OBS, or with the help of an extra plugin, so that I can display the shared buffer directly in OBS."

Thanks for sharing the qpwgraph and helvum, can I use these 2 programs to display a shared buffer in OBS?

**Reply by stephematician** [reply]
gpwgraph and helvum are mostly for audio filter graphs and won't help with displaying a shared buffer in OBS.

As far as I can tell, OBS does not have the capability to use pipewire's DMA-BUF API to capture and display video.

I believe OBS can create textures from DMA-BUFs on Linux ... but this might not be what you're after. If it is, see:

-   https://blaztinn.gitlab.io/post/dmabuf-texture-sharing/
-   https://docs.obsproject.com/reference-libobs-graphics-graphics#c.gs_texture_create_from_dmabuf

I think the process then would be to 1. write a program that sets up the capture and gets the source DMA-BUF, then 2. write a plugin/module for OBS that provides a source that can render the texture (presumably created from the DMA-BUF).

**Reply by columbarius** [reply]
> Pipewire provides the sinks and sources, and presumably some audio routers could just do the routing for video too. But no one has built any such tools or defined any protocols to coordinate between video tools like that (ala Jack). And im not aware of anyone working on that.

> I think there are thoughts that implementing camera capture can functionally serve as the protocol for general program to program sharing of video. Where programs just pretend they are "cameras". And there are people working on making cameras work on pipewire.

There are two kinds of pipewire video sinks that are interesting: context sinks (camera and screencast), which connect to PipeWire via the portal and request access to a specific node and a general sink, which will behave as a jack node, just for video and can be managed with qpwgraph. (Note: PipeWire can't mix, so only one source per input port and it won't do format conversions, this is currently worked on.)

**Reply by stephematician** [reply]
Thanks. I presume by "connect to PipeWire via the portal" you mean xdg-desktop-portal (i.e. https://docs.pipewire.org/page_portal.html)?

**Comment by stephematician** [comment]
I have some questions about the progress made on PipeWire support.

1.  Do PRs #9103 and #8153 address the constraints for the virtual camera, webcam/device capture and audio capture that were mentioned above?
2.  It isn't clear to me in this discussion if OBS supports pipewire audio capture already or not (I thought it did)?
3.  It also isn't clear to me - is OBS using pipewire screencapture on both X11 and Wayland?
4.  Are there branches/forks being used to develop pipewire video sources?

Thanks for any clarification provided.

**Reply by dimtpap** [reply]
> 1.  Do PRs #9103 and #8153 address the constraints for the virtual camera, webcam/device capture and audio capture that were mentioned above?

Partially, yes. There will be at least one more part.
> 2.  It isn't clear to me in this discussion if OBS supports pipewire audio capture already or not (I thought it did)?

PipeWire's compatibility layer, pipewire-pulse, makes OBS's PulseAudio stuff work on PipeWire.

**Reply by stephematician** [reply]
> PipeWire's compatibility layer, pipewire-pulse, makes OBS's PulseAudio stuff work on PipeWire.

Oh, right, that makes sense. So it isn't using pipewire API directly, got it.

I noticed most branches on @dimtpap, @columbarius, and @GeorgesStavracas haven't been updated in a year or more; where is the most recent work on pipewire (in particular video capture) held?

**Comment by danir-de** [comment]
Thank you for your work! Is there an estimate, when MJPEG and H264 cameras will also be supported or whats holding that back?
[The MR #9771 _(Camera portal (feat. PipeWire))_](https://github.com/obsproject/obs-studio/pull/9771) currently [only includes YUY2](https://github.com/obsproject/obs-studio/commit/364323e445645aa0e1d04547921b5824c7b4c2a5#diff-1f2f8ecc35a04162b68268688ff070725d7ae33238405ed1ed3d7b3cabb86cd0) _([see discussion comment](https://github.com/obsproject/obs-studio/pull/9771#issuecomment-1858987828))_, but is still missing these essentials codecs.

Elgato cameras (like the Elgato Streamcam and Streamcam Pro, possibly also Elgato Capture cards) only support H264 and MJPEG.

**Reply by GeorgesStavracas** [reply]
A patch exists, but it stopped working after one of the various refactorings. It needs someone to sit down and figure it out. There's no time estimation for that to happen.

**Reply by dimtpap** [reply]
https://github.com/obsproject/obs-studio/pull/11371

**Comment by westurner** [comment]
Not quite pipewire, but fast video capture and encoding APIs:
- https://github.com/ammen99/wf-recorder
- https://obsproject.com/forum/threads/no-screen-capture-option-wayland.178447/#post-655866 :
  - https://hg.sr.ht/~scoopta/wlrobs
  - https://github.com/emersion/xdg-desktop-portal-wlr
  - https://github.com/emersion/xdg-desktop-portal-wlr/wiki/%22It-doesn't-work%22-Troubleshooting-Checklist 
- https://github.com/fzwoch/obs-gnome-screencast
- moonlight protocol by CWU:
  - https://github.com/moonlight-stream works with Intel GPUs, NVIDIA RTX GPUs at 120 FPS
  - https://github.com/LizardByte/Sunshine
    > AMD: VCE 1.0 (Recommended: at least AMD VCE 3.1, HDR: 3.4)
    > Intel: VAAPI (Recommended: at least Intel HD Graphics 510 or higher, HDR: 730)
    > Nvidia: NVENC (Recommended: at least a GTX 1080)
  - https://github.com/hgaiser/moonshine
    > Requirements: NVIDIA GPU. Moonshine uses NvFBC to capture the desktop and NVENC for video encoding, both are NVIDIA specific libraries and require an NVIDIA GPU. The goal is to support more hardware in the future, while maintaining a single and therefore simple pipeline. See the todo's at the bottom for more information.
    > [...]
    > - [ ] Replace NvFBC with DRM-KMS for hardware agnostic frame capture (however at the time of writing it seems NVIDIA cards do not support this through their proprietary NVIDIA driver).
    > - [ ] Replace NVENC with [Vulkan Video Extensions](https://www.khronos.org/blog/khronos-finalizes-vulkan-video-extensions-for-accelerated-h.264-and-h.265-encode). This only really makes sense if NvFBC is replaced as well, otherwise there is still a vendor lock-in.
    - https://github.com/KhronosGroup/Vulkan-Docs/commit/542718ca30aef2c3c5afbca41de6424bd37e1808 ; VK_KHR_video_encode_queue, VK_KHR_video_encode_h264, and VK_KHR_video_encode_h265

**Comment by Sunie83** [comment]
Interop datapoint for anyone tracking this: Screen Capture (PipeWire) renders black on COSMIC (portal-cosmic 1.6.0) across OBS 32.0.4–32.2.2, both sources, normal and llvmpipe/SHM-only rendering, stream in streaming (error: none), no errors. Same OBS build + stack captures correctly under Hyprland on the same machine. Logs and full diagnosis: [pop-os/xdg-desktop-portal-cosmic#367](https://github.com/pop-os/xdg-desktop-portal-cosmic/issues/367).
